### PR TITLE
go {livesql,sqlgen,internal}: support []byte arguments for live queries

### PR DESCRIPTION
**Pull request by berfarah**
## Problem

We have UUIDs which use or convert to `[]byte`. Both livesql and the batcher rely on comparable/hashable values (values that are put into a `map`) to work and panic otherwise. This means that we can't use those custom types with livesql or batching.

### LiveSQL
The `ToArray` function has been used for creating hashable/comparable arguments in the past. Renamed it to be more explicit, and also handle `[]byte` arguments, converting them into strings.

The primary use-case for this PR is to support UUIDs. We also get support for anything else that turns into `[]byte`.

The reason we want to serialize our arguments/filters is because (unfortunately) we support type aliases in our filters, and therefore allow you to query a `int64` with `int`/`int8`/etc. In order to maintain backward compatibility there, we have to assume that we're dealing with a SQL value for caching/rerunning.

### Batching

TL;DR: do nothing

Batching deals with go struct types (not go sql types), which means we have many possible permutations of types. We _could_ take our filter values and convert them to their sql values, using that to compare. This would take more time than simply using a `[16]byte` for our UUIDs.

It would properly implement 100% compatibility with custom types. However, I have some concerns about performance in doing so, and given deadline constraints, we can add that on later rather than doing so now.

**Slices in filters will continue not to be supported for batching.**

### CODE REVIEW

File: livesql/integration_test.go [hunk 1]
```diff
@@ -90,3 +90,76 @@ func TestIntegrationBasic(t *testing.T) {
...
+	result, err := db.InsertRow(context.Background(), &User{Name: name, Uuid: uuid, Mood: &mood})
+	assert.NoError(t, err)
+	userId, err := result.LastInsertId()
+	assert.NoError(t, err)
+
+	// We use a channel to pass around query updates for testing.
+	users := make(chan *User)
+	rerunner := reactive.NewRerunner(context.Background(), func(ctx context.Context) (interface{}, error) {
+		user := &User{}
+		if err := db.QueryRow(ctx, &user, sqlgen.Filter{"mood": &mood}, nil); err != nil {
```
Comment (berfarah) [inline livesql/integration_test.go]: This means UUIDs should definitely work

File: livesql/integration_test.go [hunk 2]
```diff
@@ -90,3 +90,76 @@ func TestIntegrationBasic(t *testing.T) {
...
+	newUuid := testfixtures.CustomTypeFromString("11232481203")
+
+	result, err := db.InsertRow(context.Background(), &User{Name: name, Uuid: uuid, Mood: &mood})
+	assert.NoError(t, err)
+	userId, err := result.LastInsertId()
+	assert.NoError(t, err)
+
+	// We use a channel to pass around query updates for testing.
+	users := make(chan *User)
+	rerunner := reactive.NewRerunner(context.Background(), func(ctx context.Context) (interface{}, error) {
```
Comment (changpingc) [inline livesql/integration_test.go]: try using batch.WithBatching(ctx)

**Comment by changpingc** [comment]
cute idea: check reflect.Type.Comparable()

File: internal/reflect.go [hunk 3]
```diff
@@ -31,10 +31,17 @@ func init() {
...
 }
 
-// ToArray converts a []interface{} slice into an equivalent fixed-length array
+// MakeHashable converts a []interface{} slice into an equivalent fixed-length array
 // [...]interface{} for use as a comparable map key
-//
-func ToArray(s []interface{}) interface{} {
+func MakeHashable(s []interface{}) interface{} {
+	// Convert byte slices into strings as they are otherwise not comparable/hashable.
+	for i, elem := range s {
```
[the remark this answers is not in the document]
Comment (berfarah) [reply internal/reflect.go]: Hmm, I think I didn't quite get pass by value here - I'll change this so it makes a new slice and check some benchmarks.

File: internal/reflect.go [hunk 4]
```diff
@@ -31,10 +31,17 @@ func init() {
 	interfaceTyp = reflect.TypeOf(&x).Elem()
 }
 
-// ToArray converts a []interface{} slice into an equivalent fixed-length array
+// MakeHashable converts a []interface{} slice into an equivalent fixed-length array
 // [...]interface{} for use as a comparable map key
-//
-func ToArray(s []interface{}) interface{} {
+func MakeHashable(s []interface{}) interface{} {
```
[the remark this answers is not in the document]
Comment (berfarah) [reply internal/reflect.go]: My concern would be breaking things at the SQL layer if I put it higher up. If I were to pick out a refactor that could be worth doing, it's making the filter more responsible for its various conversions so that we don't have to hold in our heads what these things ^ are used for.

[the remark this answers is not in the document]
Comment (berfarah) [reply internal/reflect.go]: Looks like batching doesn't actually use the `driver.Value`. Need to find a way to do so.

**Comment by berfarah** [comment]
@stephen for batching, right now the code compares non-sql-value types, which means identifiers would have to be comparable. This means that without changing our uuids from `[]byte` to `[16]byte`, batching is broken for UUIDs. I think what I want to do is go with the UUID change for now and take more time to consider our various (de)serialization layers.

**Comment by berfarah** [comment]
@stephen I'm ready to merge this as soon as I have a 👍

File: internal/reflect.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (berfarah) [reply internal/reflect.go]: This no longer modifies the original slice

File: internal/reflect.go [hunk 5]
```diff
@@ -31,34 +31,44 @@ func init() {
...
+	}
+
+	// Return arrays as they are comparable/hashable.
+	switch len(d) {
 	// fast code paths for short arrays:
 	case 0:
 		return [...]interface{}{}
 	case 1:
-		return [...]interface{}{s[0]}
+		return [...]interface{}{d[0]}
```
Comment (stephen) [inline internal/reflect.go]: super nit: consider changing out `s` -> `d` in the arguments instead so that these references stay the same.

Comment (stephen) [reply internal/reflect.go]: (i also checked _find all references_ so it seems fine to leave)

File: internal/reflect.go [hunk 6]
```diff
@@ -31,34 +31,44 @@ func init() {
 	interfaceTyp = reflect.TypeOf(&x).Elem()
 }
 
-// ToArray converts a []interface{} slice into an equivalent fixed-length array
+// MakeHashable converts a []interface{} slice into an equivalent fixed-length array
```
Comment (stephen) [inline internal/reflect.go]: nit: consider doing the rename in a separate step to make the semantic change easier to read

Comment (berfarah) [reply internal/reflect.go]: Done
